### PR TITLE
Add previous data on transform

### LINKED ISSUE #7054: [Feature Request] Making previousData accessible from save and transform method.
**Is your feature request related to a problem? Please describe.**

It may be better to access previous data state(to be exact, before saved state) from save and transform method than it is only accessible from dataProvider.
Now, as I know with reading code useEditController and react-admin document, DataProvider is only way (except for customizing default component) to access previousData, but as more resource depended logics are implemented though intrinsically they are loosely coupled, it is fat soon. 

If it can be accessed from transform and save methods,  it's convenient for limiting scope narrower and  modifying request parameter logic form layer can be completed, in the result, we will no longer implement conditional logic using previousData in dataProvider , then more agnostic feature against resource type.

**Describe the solution you'd like**
Add more the previousData argument and assign indirect  record 
[this line](https://github.com/marmelab/react-admin/blob/43c4fafc8bcedebc386c7d3dc3b63cfd56420a17/packages/ra-core/src/controller/details/useEditController.ts#L176)
and pass record value second parameter [https://github.com/marmelab/react-admin/blob/43c4fafc8bcedebc386c7d3dc3b63cfd56420a17/packages/ra-core/src/controller/details/useEditController.ts#L178](this code).

**Describe alternatives you've considered**

I read [https://marmelab.com/react-admin/CreateEdit.html#usecreatecontroller-and-useeditcontroller](this section of document).
Edit props completely rewritable and provide alternative controller props to store state of previous  data. So, if this features implements not desired for framework, we can make this feature.

**Additional context**

### PR DESCRIPTION
**Pull request by tkow**
Closes #7054 

Add previous data argument transform method. My tests added is redundant because previousData always is passed and  asserted in every tests but, it's noting worth to keep description about expecting behavior, ready for this behavior is modified (however if you don't need it, I'll remove them).

### CODE REVIEW

File: packages/ra-core/src/dataProvider/useUpdate.ts [hunk 1]
```diff
@@ -382,7 +382,7 @@ export const useUpdate = <RecordType extends RaRecord = any>(
         }
     };
 
-    return [update, mutation];
```
Comment (djhi) [inline packages/ra-core/src/dataProvider/useUpdate.ts:385]: You don't need to pass refs in dependencies

**Review by djhi — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/dataProvider/useUpdate.ts (hunk 1 above)
Comment (tkow) [reply packages/ra-core/src/dataProvider/useUpdate.ts:385]: @djhi 
Sorry, I marked draft and I intended to revise but, I couldn't figure out how useUpdate cache previousData.
Now, I understand lifecycle in useQuery makes record changed and form's record nothing relevant, so at least one changes by getOne, so we can use record variable.
But, I thought this is very confused so, I make the variable with clearer name as cachedData. I removed draft sign. So could you review again?

File: docs/CreateEdit.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ title: "The Create and Edit Views"
 
 # The Create and Edit Views
 
-`<Resource>` maps URLs to components - it takes care of *routing*. When you set a component as the `create` prop for a Resource, react-admin renders that component when users go to the `/[resource]/create` URL. When you set a component as the `edit` prop for a resource, react-admin renders that component when users go to the `/[resource]/:id` URL. 
+`<Resource>` maps URLs to components - it takes care of *routing*. When you set a component as the `create` prop for a Resource, react-admin renders that component when users go to the `/[resource]/create` URL. When you set a component as the `edit` prop for a resource, react-admin renders that component when users go to the `/[resource]/:id` URL.
```
Comment (djhi) [inline docs/CreateEdit.md:8]: There are a lot of unrelated and unnecessary changes here due to your markdown formatter probably. Can you please revert them?

File: docs/CreateEdit.md [hunk 3]
```diff
@@ -541,6 +541,27 @@ The `transform` function can also return a `Promise`, which allows you to do all
 
 **Tip**: The message will be translated.
 
+**Tip**: `<Edit>`'s transform prop can recieve below variables from second argument.
+
+|  field         | Description                                                                                |
+| -------------- | ------------------------------------------------------------------------------------------ |
+| `previousData` | Previous data fetched or last updated on `dataProvider.update`                             |
+
```
Comment (djhi) [inline docs/CreateEdit.md:549]:
```suggestion
**Tip**: `<Edit>`'s transform prop function also get the `previousData` in its second argument:
```

File: docs/CreateEdit.md [hunk 4]
```diff
@@ -2289,6 +2310,37 @@ const dataProvider = {
 }
 ```
 
+**Tip**: `<SaveButton>`'s transform prop inside `<Edit>` can recieve below variables from second argument.
+
+|  field         | Description                                                                                |
+| -------------- | ------------------------------------------------------------------------------------------ |
+| `previousData` | Previous data fetched or last updated on `dataProvider.update`                             |
```
Comment (djhi) [inline docs/CreateEdit.md:2317]:
```suggestion
**Tip**: `<Edit>`'s transform prop function also get the `previousData` in its second argument:
```

**Review by djhi — CHANGES_REQUESTED** [review]

**Comment by tkow** [comment]
@djhi 
Now, I revised them, and please review again.

`Vercel – react-admin-storybook`'s deploy is failed. Is that no problem?

**Comment by tkow** [comment]
@djhi I'm sorry, but I couldn't find why Vercel - react-admin-storybook deployment has failed because I have no idea to know without vercel logs. It can be build in my local environment. 
Could you help me when you have time?

File: packages/ra-core/src/controller/edit/useEditController.ts [hunk 5]
```diff
@@ -116,9 +118,13 @@ export const useEditController = <RecordType extends RaRecord = any>(
         ) =>
             Promise.resolve(
                 transformFromSave
-                    ? transformFromSave(data)
+                    ? transformFromSave(data, {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/edit/useEditController.ts:121]: check the linter warning for missing dependency

File: docs/CreateEdit.md [hunk 6]
```diff
@@ -542,6 +542,22 @@ The `transform` function can also return a `Promise`, which allows you to do all
 
 **Tip**: The message will be translated.
 
+**Tip**: `<Edit>`'s transform prop function also get the `previousData` in its second argument:
+
+```jsx
+export const UserEdit = (props) => {
+    const transform = (data, { previousData }) => ({
```
Comment (fzaninotto) [inline docs/CreateEdit.md:549]: I'd prefer that the transform function signature be `(data, previousData)`

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Comment by fzaninotto** [comment]
The react-admin-storybook deployment randomly fails, don't worry about It. I redeployed it manually.

File: docs/CreateEdit.md (hunk 6 above)
Comment (tkow) [reply docs/CreateEdit.md:549]: @fzaninotto 

Sorry, but I'd prefer my way, because it rarely  breaks user code when we need to add more optionalData compared to push it to the tail of arguments though the change may be seldom.
Increasing argument have risks when we want to add new argument, intermediate arguments can't change order without breaking change and even if they are not used in some cases, we may need to assign undefined when we call  or when we set no used variables' specification (like, (data, _, someEditControllerParameter) as event hander.

Argument addition as option is better way for me if the argument is really optional value. In this case, users can ignore previousData so, it's optional value.
But, I can't say that previousData must be optional form because I haven't found that we need more additional parameter in the future. My point is opposite to YAGNI, and  previousData may be just enough parameter in all edit controller context parameters to satisfy user demands. So, I'll fix that  if you still want to do nonetheless you have read this comment.

File: packages/ra-core/src/controller/edit/useEditController.ts (hunk 5 above)
Comment (tkow) [reply packages/ra-core/src/controller/edit/useEditController.ts:121]: Thank you I've not noticed. It's green now.

**Review by CryptoGodfatherVA2 — APPROVED** [review]
VAR5

**Review by CryptoGodfatherVA2 — APPROVED** [review]

**Comment by fzaninotto** [comment]
we'll wait for #7087 to be merged to merge this one, as I prefer that you rebase your small PR rather than asking @djhi to rebase his mega-jumbo-huge-not-so-small PR.

**Comment by tkow** [comment]
I rebased my commits to remove no diff next branch merge commits.

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!
